### PR TITLE
Use upgrade callback to pick the correct carousel class.

### PR DESCRIPTION
**Pull request by camelburrito**
fixes - #5828 

FYI : js bin showing how constructor is over written when we return a class in the constructor 

http://jsbin.com/wohiruq/

### CODE REVIEW

**Comment by camelburrito** [comment]
CC @dvoytenko, @jridgewell

File: extensions/amp-carousel/0.1/amp-carousel.js [hunk 1]
```diff
@@ -20,18 +20,18 @@ import {AmpScrollableCarousel} from './scrollable-carousel';
...
 
-class CarouselSelector {
-
-  constructor(element) {
-    if (element.getAttribute('type') == 'slides') {
-      return new AmpSlideScroll(element);
+class CarouselSelector extends AMP.BaseElement {
+  /** @override */
+  upgradeCallback() {
+    if (this.element.getAttribute('type') == 'slides') {
```
Comment (cramforce) [inline extensions/amp-carousel/0.1/amp-carousel.js]: This isn't yet fixing the bug. The bug is that this is called before any attributes are passed.

The correct solution is to return a promise from this.

Just wrapping everything in

```JS
return waitForChildPromise(this.element, () => {
  return this.childNodes.length > 0;
}).then(/* existing code in upgradeCallback */);
```

should work.

Comment (jridgewell) [inline extensions/amp-carousel/0.1/amp-carousel.js]: Attributes aren't available to `#upgradeCallback`?

Comment (dvoytenko) [inline extensions/amp-carousel/0.1/amp-carousel.js]: The intention is that `upgradeCallback` is called only AFTER attributes are available yet. `upgradeCallback` and only after "connected". If that's not the case - it's a bug and simply needs to be fixed. Under no circumstances, however, we should wait for children.

**Pull request merged**